File: packages/gavio-py/gavio/interceptors/pii/scanners/routing_number.py
"""US ABA routing-number scanner — 9 digits + mod-10 checksum (FinTech pack)."""

from __future__ import annotations

import re

from ..context import ScanContext
from ..match import PiiMatch
from ..scanner import PiiScanner

_ROUTING = re.compile(r"\b\d{9}\b")
_WEIGHTS = (3, 7, 1, 3, 7, 1, 3, 7, 1)


def valid_routing_number(candidate: str) -> bool:
    """ABA checksum: weighted digit sum must be a non-zero multiple of 10."""
    if len(candidate) != 9 or not candidate.isdigit():
        return False
    total = sum(w * int(d) for w, d in zip(_WEIGHTS, candidate, strict=True))
    return total > 0 and total % 10 == 0


class RoutingNumberScanner(PiiScanner):
    entity_type = "ROUTING_NUMBER"

    def scan(self, text: str, ctx: ScanContext) -> list[PiiMatch]:
        out: list[PiiMatch] = []
        for m in _ROUTING.finditer(text):
            if not valid_routing_number(m.group()):
                continue
            idx = ctx.next_index(self.entity_type)
            out.append(
                PiiMatch(
                    entity_type=self.entity_type,
                    start=m.start(),
                    end=m.end(),
                    value=m.group(),
                    replacement=f"[ROUTING_NUMBER_{idx}]",
                )
            )
        return out

File: packages/gavio-js/src/interceptors/pii/scanners/routing-number.ts
/** US ABA routing-number scanner — 9 digits + mod-10 checksum (F-SEC-01, FinTech pack). */

import type { ScanContext } from '../context.js'
import { makeMatch } from '../match.js'
import type { PiiMatch } from '../match.js'
import type { PiiScanner } from '../scanner.js'

const ROUTING = /\b\d{9}\b/g
const WEIGHTS = [3, 7, 1, 3, 7, 1, 3, 7, 1]

/** ABA checksum: weighted digit sum must be a non-zero multiple of 10. */
export function validRoutingNumber(candidate: string): boolean {
  if (candidate.length !== 9) return false
  let sum = 0
  for (let i = 0; i < 9; i++) sum += WEIGHTS[i]! * (candidate.charCodeAt(i) - 48)
  return sum > 0 && sum % 10 === 0
}

export function routingNumberScanner(): PiiScanner {
  return {
    entityType: 'ROUTING_NUMBER',
    tier: 1,
    scan(text: string, ctx: ScanContext): PiiMatch[] {
      const out: PiiMatch[] = []
      for (const m of text.matchAll(ROUTING)) {
        if (!validRoutingNumber(m[0])) continue
        const idx = ctx.nextIndex('ROUTING_NUMBER')
        out.push(
          makeMatch({
            entityType: 'ROUTING_NUMBER',
            start: m.index,
            end: m.index + m[0].length,
            value: m[0],
            replacement: `[ROUTING_NUMBER_${idx}]`,
          }),
        )
      }
      return out
    },
  }
}
